src/os/bluestore/BlueStore.cc

```diff
@@ -8671,21 +8674,24 @@ void BlueStore::_deferred_submit_unlock(OpSequencer *osr)
   bdev->aio_submit(&b->ioc);
 }
 
-void BlueStore::_deferred_aio_finish(DeferredBatch *b)
+void BlueStore::_deferred_aio_finish(OpSequencer *osr)
 {
-  OpSequencer *osr = b->osr;
-  dout(10) << __func__ << " osr " << osr << " b " << b << dendl;
+  dout(10) << __func__ << " osr " << osr << dendl;
+  assert(osr->deferred_running);
+  DeferredBatch *b = osr->deferred_running;
 
   {
     std::lock_guard<std::mutex> l(deferred_lock);
-    auto p = std::find(osr->deferred_running.begin(),
-		       osr->deferred_running.end(),
-		       b);
-    assert(p != osr->deferred_running.end());
-    osr->deferred_running.erase(p);
-    if (osr->deferred_running.empty() && !osr->deferred_pending) {
+    assert(osr->deferred_running == b);
+    osr->deferred_running = nullptr;
+    if (!osr->deferred_pending) {
       auto q = deferred_queue.iterator_to(*osr);
       deferred_queue.erase(q);
+    } else if (deferred_aggressive) {
+      dout(20) << __func__ << " queuing async deferred_try_submit" << dendl;
+      finishers[0]->queue(new FunctionContext([&](int) {
```

Comment:
I'm not completely up on how the finishers work, but since this function's about to return won't the variables in the capture-by-reference closure be invalid when the lambda gets called?